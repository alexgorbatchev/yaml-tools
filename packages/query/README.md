# @yaml-tools/query

A utility for the [yaml](https://github.com/eemeli/yaml) that lets you query a YAML document using simple object syntax. This module is a simplified fork of the [yaml-doc-query](https://github.com/Xiphe/yaml-doc-query).

```ts
import yaml from '@yaml-tools/inline';
import query from '@yaml-tools/query';

const doc = yaml`
  one: hello world
  two:
    - three: 4
    - five: 5
`;

const selector = query(doc);

selector.one();
// hello world

selector.two[0].three();
// 4
```
