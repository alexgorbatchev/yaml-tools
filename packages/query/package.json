{
  "name": "@yaml-tools/query",
  "version": "1.0.3",
  "description": "A utility to query a YAML document using simple object syntax.",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+ssh://git@github.com:alexgorbatchev/yaml-tools.git"
  },
  "homepage": "https://github.com/alexgorbatchev/yaml-tools/tree/master/packages/query",
  "author": {
    "name": "Alex Gorbatchev"
  },
  "type": "module",
  "types": "./dist/index.d.ts",
  "files": ["./dist/*"],
  "exports": "./dist/index.js",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "lint": "eslint 'src'",
    "dev": "tsc --project . --watch",
    "clean": "rm -fr ./dist",
    "build": "pnpm run clean && tsc --project tsconfig.build.json",
    "test": "node --experimental-vm-modules $(pnpm root)/jest/bin/jest.js"
  },
  "dependencies": {
    "yaml": "^2.1.3"
  },
  "devDependencies": {
    "jest": "^29.3.0",
    "ts-jest": "^29.0.3",
    "typescript": "^4.9.3",
    "@yaml-tools/inline": "workspace:*"
  }
}
