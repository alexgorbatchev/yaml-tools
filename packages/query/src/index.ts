export { query as default } from './query.js';
