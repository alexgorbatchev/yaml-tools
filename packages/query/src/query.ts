import YAML from 'yaml';

export type QueryResult = {
  (): YAML.Node | undefined;
  [key: string | number | symbol]: QueryResult;
};

export function query(document: YAML.Document): QueryResult {
  return process(document.contents, document);
}

export class RangeError extends Error {
  range?: YAML.Range;

  constructor(message: string, range?: YAML.Range) {
    super(message);
    this.range = range;
    Object.setPrototypeOf(this, RangeError.prototype);
  }
}

function isValidIndexType(v: unknown): v is string | number | symbol {
  return ['string', 'number', 'symbol'].includes(typeof v);
}

function process(node: YAML.Node | null | undefined, document: YAML.Document): any {
  if (YAML.isAlias(node)) {
    node = node.resolve(document);
  }

  return new Proxy(
    () => {
      if (!node) {
        return undefined;
      }

      return node;
    },
    {
      get(_, prop): any {
        if (YAML.isMap<YAML.Node, YAML.Node>(node)) {
          const match = node.items.find(({ key }) => {
            if (!YAML.isScalar(key)) {
              throw new RangeError(`Unexpected ${key.constructor.name}`, key.range || undefined);
            }

            if (!isValidIndexType(key.value)) {
              throw new RangeError(`Unexpected ${typeof key.value}`, key.range || undefined);
            }

            return (typeof key.value === 'number' ? String(key.value) : key.value) === prop;
          });

          return process(match?.value, document);
        }

        if (YAML.isSeq<YAML.Node>(node)) {
          return process(node.items[prop as any], document);
        }

        if (!node) {
          return process(node, document);
        }
      },
    },
  );
}
