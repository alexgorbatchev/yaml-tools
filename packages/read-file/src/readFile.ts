import visitOperator from '@yaml-tools/visit-operator';
import fs from 'node:fs';
import { dirname, join } from 'node:path/posix';
import YAML from 'yaml';

interface Options {
  /**
   * Current working directory, defaults to `process.cwd()`. This option only
   * has effect on the `+include` operator, more specifically on the `~/` paths
   * used with it.
   */
  cwd?: string;
}

/**
 * Reads a YAML file while processing `+include: ./path/to.yaml` nodes merging
 * contents of multiple YAML files.
 *
 * Include file path must be either relative to the YAML file and start with a
 * dot `./foo` or relative to the `opts.cwd` and start with a tilde `~/foo`.
 * `opts.cwd` defaults to `process.cwd()`.
 *
 * This function is syncronous because `YAML.visit()` is syncronous.
 */
export const readFile = (fileName: string, opts: Options = {}): YAML.Document => {
  // this method has to be syncronous because YAML.visit is syncronous

  const cwd = opts?.cwd ?? process.cwd();
  const content = fs.readFileSync(fileName, 'utf-8');
  const context = dirname(fileName);

  return visitOperator(YAML.parseDocument(content), '+include', (_, node) => {
    if (!YAML.isScalar(node.value) || typeof node.value.value !== 'string') {
      throw new Error('+include value must be a string');
    }

    const includePath = node.value.value;
    const includeFileName = includePath.startsWith('~/')
      ? includePath.replace(/^~\//, cwd + '/')
      : join(context, includePath);

    return readFile(includeFileName, { cwd }).contents as any;
  });
};
