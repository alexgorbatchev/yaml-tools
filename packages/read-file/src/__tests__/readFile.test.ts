import query from '@yaml-tools/query';
import { join } from 'path';

import { hasFilePath, readFile, withFilePath } from '../readFile.js';
import { $dirname } from './helpers.js';

const __dirname = $dirname(import.meta);
const cwd = join(__dirname, 'fixtures');
const fixturePath = join(__dirname, 'fixtures/subfolder/readFile.yaml');

describe('readFile', () => {
  test('processes +include', () => {
    const yaml = readFile(fixturePath, { cwd });

    expect(yaml.toString()).toMatchInlineSnapshot(`
      "anchors:
        map:
          this:
            is:
              - deep:
                  - $types/node: ^16.11.6
                  - eslint: ../local/path
                  - tslib: next
        sequence:
          - 0
          - a
          - b
          - c
          - 1
        another: value
        boo: 10
      "
    `);
  });

  test('does not process +include', () => {
    const yaml = readFile(fixturePath, { cwd, disableIncludes: true });

    expect(yaml.toString()).toMatchInlineSnapshot(`
      "anchors:
        map:
          +include: ~/map.yaml
        sequence:
          - 0
          - +include: ../sequence.yaml
          - 1
        another: value
        boo: 10
      "
    `);
  });

  test('adds `filePath`', () => {
    const yaml = readFile(fixturePath, { cwd });
    const selector = query(yaml);
    const getFilePath = (v: any) => withFilePath(v)?.filePath.replace(`${cwd}/`, '');

    expect(getFilePath(yaml.contents)).toBe('subfolder/readFile.yaml');
    expect(getFilePath(selector.anchors.map.this.is())).toBe(`map.yaml`);
    expect(getFilePath(selector.anchors.sequence[1]())).toBe(`sequence.yaml`);
  });

  test('does not add `filePath`', () => {
    const yaml = readFile(fixturePath, { cwd, disableFilePathInjection: true });
    expect(hasFilePath(yaml.contents!)).toBe(false);
  });
});
