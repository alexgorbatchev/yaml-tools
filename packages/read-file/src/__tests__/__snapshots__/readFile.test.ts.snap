// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readFile 1`] = `
"anchors:
  map:
    this:
      is:
        - deep:
            - $types/node: ^16.11.6
            - eslint: ../local/path
            - tslib: next
  sequence:
    - 0
    - a
    - b
    - c
    - 1
  another: value
  boo: 10
"
`;
